fix: make query results API resource db_role-aware

ajkerrigan opened the pull request:
[ENG-5796](https://stacklet.atlassian.net/browse/ENG-5796)

### what

Make sure that the latest query results are specific to the current user's `db_role`, for GET requests to the direct query result endpoint:

```console
https://redash.canary.stacklet.io/api/queries/1/results.json
```

### why

Aegon encountered an issue where users with limited access were running queries, and then admins were seeing those limited cached results instead of the full result set. #77 took care of the UI driven flow from one report, but didn't address this request flow that is common for referencing cached query results in policies with robot user API keys.

### testing

Reproduced in my sandbox to verify the change interactively, and confirmed that the included unit test passes now but failed without the change. 

### docs

n/a

[ENG-5796]: https://stacklet.atlassian.net/browse/ENG-5796?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Ensure GET `api/queries/:id/results.json` returns the latest cached result for the current user's `db_role`, not the query's global `latest_query_data`.
> 
> - **Backend**:
>   - **`QueryResultResource.get`**: When fetching by `query_id` (without `query_result_id`), ignore `query.latest_query_data_id` and use `models.QueryResult.get_latest` with `db_role=getattr(self.current_user, "db_role", None)` to retrieve role-specific cached results.
> - **Tests**:
>   - Add `test_query_results_by_db_role` validating that admin and limited users receive their respective cached results for the same query.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a19b166099ea2c7399e7d4a01a23525ed3e603ce. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

ajkerrigan committed:
fix: make /api/queries/{id}/results db_role-aware

ajkerrigan committed:
test query result api responses by db role

johnsca commented on redash/handlers/query_results.py:
```diff
@@ -318,11 +318,16 @@ def get(self, query_id=None, query_result_id=None, filetype="json"):
+                    max_age=-1,
+                    is_hash=True,
+                    db_role=getattr(self.current_user, "db_role", None),
                 )
```
You still need the `get_object_or_404` wrapper, but I think it should be straightforward. Something like:
```suggestion
                query_result = get_object_or_404(
                    models.QueryResult.get_latest,
                    data_source=query.data_source,
                    query=query.query_hash,
                    max_age=-1,
                    is_hash=True,
                    db_role=getattr(self.current_user, "db_role", None),
                )
```

I'm a little surprised the existing tests don't include a 404 case. Either way it's probably worth adding one for the user / admin case (user with only an admin result, admin with only a user result).

ajkerrigan replied:
Hmm I actually like the user experience of this better without the `get_object_or_404()`. :eyes: 

With the change in this PR, we get an empty query result which aborts [later in the same method](https://github.com/stacklet/redash/blob/a19b166099ea2c7399e7d4a01a23525ed3e603ce/redash/handlers/query_results.py#L382-L383) with a 404 status code and the message:

> No cached result found for this query.

If we add the `get_object_or_404()` wrapper, the caller still gets a 404 response but with a more generic message:

> The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again.

I _think_ the more specific response message is useful for what we're doing here, though it's not a super strong opinion :thinking: . Does this context affect your thoughts one way or the other?

ajkerrigan replied:
Meh nevermind. I'm outnumbered here by you and Cursor :sweat_smile: , and didn't have a particularly strong aversion to the wrapper anyway. Just added it.

johnsca replied:
Ah, I missed that it was handled in that other bit and the wrapper does make that handling a bit redundant, but since it was already like that, I guess it's better to keep the existing behavior? 🤷

johnsca reviewed (changes requested):
In addition to the request below, I also noticed that [run_query calls create_tables_from_query_ids](https://github.com/stacklet/redash/blob/4b935cd9062b5b093be3cf6ab8fb905d6e61888d/redash/query_runner/query_results.py#L176), which [calls get_query_results with bring_from_cache=True](https://github.com/stacklet/redash/blob/4b935cd9062b5b093be3cf6ab8fb905d6e61888d/redash/query_runner/query_results.py#L87), which [also uses latest_query_data_id (and latest_query_data)](https://github.com/stacklet/redash/blob/4b935cd9062b5b093be3cf6ab8fb905d6e61888d/redash/query_runner/query_results.py#L69-L70) and thus suffers from this same issue.

It looks like this is used for queries against a QueryResults data source, so would manifest when using that. The query results for the given queries are mapped back into an in-memory SQLite DB and the queries run against that; those in-memory pseudo-tables won't have RLS, so it is important that we properly filter the cached query results to the ones that the user has access to. I think that's mostly already covered except for the admin case that you're fixing here, so if you could please apply this fix to the `bring_from_cache` case in `get_query_results` as well, that would be great.

ajkerrigan committed:
fix: enforce db_role for the query results data source

When referencing a cached query using the query results data source (QRDS),
fetch results specific to the caller's db_role.

Note that this will only apply to new query runs.

ajkerrigan committed:
bake a db_role into saved query_results runs

ajkerrigan commented:
> In addition to the request below, I also noticed that [run_query calls create_tables_from_query_ids](https://github.com/stacklet/redash/blob/4b935cd9062b5b093be3cf6ab8fb905d6e61888d/redash/query_runner/query_results.py#L176), which [calls get_query_results with bring_from_cache=True](https://github.com/stacklet/redash/blob/4b935cd9062b5b093be3cf6ab8fb905d6e61888d/redash/query_runner/query_results.py#L87), which [also uses latest_query_data_id (and latest_query_data)](https://github.com/stacklet/redash/blob/4b935cd9062b5b093be3cf6ab8fb905d6e61888d/redash/query_runner/query_results.py#L69-L70) and thus suffers from this same issue.
> 
> It looks like this is used for queries against a QueryResults data source, so would manifest when using that. The query results for the given queries are mapped back into an in-memory SQLite DB and the queries run against that; those in-memory pseudo-tables won't have RLS, so it is important that we properly filter the cached query results to the ones that the user has access to. I think that's mostly already covered except for the admin case that you're fixing here, so if you could please apply this fix to the `bring_from_cache` case in `get_query_results` as well, that would be great.

Very good catch there, good eye @johnsca :monocle_face: . That definitely makes some loopy code paths even trickier. One thing I noticed was that if I ran a query as an admin, I could write a new query against the Query Results Data Source as a limited user and get cached admin results from the original query... but that derived query's result record would still show the limited user's db_role. Once we fix the main issue, there's still no obvious db_role context for the cached query references.

So I've made a change [here](https://github.com/stacklet/redash/blob/750ae8a9d53e4f182331b22af8e675b15bbb75b4/redash/query_runner/query_results.py#L220) purely so we start recording that context. It's not clear to me if we should do something like:

- Purge all saved query results associated with the query results data source once we land this
- Use that "second tier" db_role as an extra check when looking up existing results from QRDS queries
- Just keep this in mind as a consideration related to tech debt cleanup in [ENG-5812](https://stacklet.atlassian.net/browse/ENG-5812)

[ENG-5812]: https://stacklet.atlassian.net/browse/ENG-5812?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ajkerrigan committed:
ok, ok... include the 404 wrapper

ajkerrigan commented:
FWIW I've also added a wiki section [here](https://stacklet.atlassian.net/wiki/spaces/ENG/pages/776339461/Redash+Installation#Debug-Worker-Processes) with some info about debugging worker processes. Since those cached_query lookups were happening in the background worker rather than on the server process.
